File: strategies/1-brute-force/TSPNaive.java
/**
 * Brute Force Strategy: Traveling Salesman Problem (TSP) - Naive Approach
 * Core Idea: Try all possible permutations of cities to find the shortest route
 * Time Complexity: O(n!) where n is the number of cities
 * Space Complexity: O(n) for recursion stack
 */
public class TSPNaive {
    
    /**
     * Calculate the total distance for a given route
     * @param graph Adjacency matrix representing distances between cities
     * @param route Array representing the order of cities to visit
     * @return Total distance of the route
     */
    public static int calculateRouteDistance(int[][] graph, int[] route) {
        int totalDistance = 0;
        int n = route.length;
        
        // Calculate distance from each city to the next
        for (int i = 0; i < n - 1; i++) {
            totalDistance += graph[route[i]][route[i + 1]];
        }
        
        // Add distance from last city back to starting city
        totalDistance += graph[route[n - 1]][route[0]];
        
        return totalDistance;
    }
    
    /**
     * Solve TSP using brute force approach
     * @param graph Adjacency matrix representing distances between cities
     * @return Minimum distance for the shortest route
     */
    public static TSPResult solveTSP(int[][] graph) {
        int n = graph.length;
        if (n <= 1) return new TSPResult(0, new int[]{0});
        
        // Create initial route starting from city 0
        int[] cities = new int[n];
        for (int i = 0; i < n; i++) {
            cities[i] = i;
        }
        
        int minDistance = Integer.MAX_VALUE;
        int[] bestRoute = new int[n];
        
        // Try all possible permutations of cities (except starting city)
        do {
            int currentDistance = calculateRouteDistance(graph, cities);
            if (currentDistance < minDistance) {
                minDistance = currentDistance;
                System.arraycopy(cities, 0, bestRoute, 0, n);
            }
        } while (nextPermutation(cities, 1)); // Start from index 1 to keep city 0 as start
        
        return new TSPResult(minDistance, bestRoute);
    }
    
    /**
     * Generate next lexicographic permutation
     * @param arr Array to permute
     * @param start Starting index for permutation
     * @return true if next permutation exists, false otherwise
     */
    private static boolean nextPermutation(int[] arr, int start) {
        // Find the largest index i such that arr[i] < arr[i + 1]
        int i = arr.length - 2;
        while (i >= start && arr[i] >= arr[i + 1]) {
            i--;
        }
        
        if (i < start) return false; // No more permutations
        
        // Find the largest index j such that arr[i] < arr[j]
        int j = arr.length - 1;
        while (arr[j] <= arr[i]) {
            j--;
        }
        
        // Swap arr[i] and arr[j]
        swap(arr, i, j);
        
        // Reverse the suffix starting at arr[i + 1]
        reverse(arr, i + 1);
        
        return true;
    }
    
    private static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
    
    private static void reverse(int[] arr, int start) {
        int end = arr.length - 1;
        while (start < end) {
            swap(arr, start, end);
            start++;
            end--;
        }
    }
    
    /**
     * Helper class to store TSP result
     */
    static class TSPResult {
        int minDistance;
        int[] bestRoute;
        
        TSPResult(int distance, int[] route) {
            this.minDistance = distance;
            this.bestRoute = route;
        }
    }
    
    /**
     * Helper method to print route
     */
    public static void printRoute(int[] route) {
        System.out.print("Route: ");
        for (int i = 0; i < route.length; i++) {
            System.out.print(route[i]);
            if (i < route.length - 1) System.out.print(" -> ");
        }
        System.out.println(" -> " + route[0]); // Return to start
    }
    
    public static void main(String[] args) {
        System.out.println("=== Traveling Salesman Problem (TSP) - Brute Force ===");
        
        // Test Case 1: Simple 4-city problem
        System.out.println("Test Case 1: 4 cities");
        int[][] graph1 = {
            {0, 10, 15, 20},
            {10, 0, 35, 25},
            {15, 35, 0, 30},
            {20, 25, 30, 0}
        };
        
        System.out.println("Distance matrix:");
        printMatrix(graph1);
        
        TSPResult result1 = solveTSP(graph1);
        System.out.println("Minimum distance: " + result1.minDistance);
        printRoute(result1.bestRoute);
        System.out.println();
        
        // Test Case 2: 3-city triangle
        System.out.println("Test Case 2: 3 cities");
        int[][] graph2 = {
            {0, 5, 8},
            {5, 0, 3},
            {8, 3, 0}
        };
        
        System.out.println("Distance matrix:");
        printMatrix(graph2);
        
        TSPResult result2 = solveTSP(graph2);
        System.out.println("Minimum distance: " + result2.minDistance);
        printRoute(result2.bestRoute);
        System.out.println();
        
        // Test Case 3: 2-city problem (trivial)
        System.out.println("Test Case 3: 2 cities");
        int[][] graph3 = {
            {0, 7},
            {7, 0}
        };
        
        System.out.println("Distance matrix:");
        printMatrix(graph3);
        
        TSPResult result3 = solveTSP(graph3);
        System.out.println("Minimum distance: " + result3.minDistance);
        printRoute(result3.bestRoute);
    }
    
    private static void printMatrix(int[][] matrix) {
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[i].length; j++) {
                System.out.printf("%3d ", matrix[i][j]);
            }
            System.out.println();
        }
    }
}

File: strategies/1-brute-force/TSPNaive.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <string.h>

/**
 * Brute Force Strategy: Traveling Salesman Problem (TSP) - Naive Approach
 * Core Idea: Try all possible permutations of cities to find the shortest route
 * Time Complexity: O(n!) where n is the number of cities
 * Space Complexity: O(n) for recursion stack
 */

#define MAX_CITIES 10

/**
 * Calculate the total distance for a given route
 * @param graph Adjacency matrix representing distances between cities
 * @param route Array representing the order of cities to visit
 * @param n Number of cities
 * @return Total distance of the route
 */
int calculateRouteDistance(int graph[][MAX_CITIES], int route[], int n) {
    int totalDistance = 0;
    
    // Calculate distance from each city to the next
    for (int i = 0; i < n - 1; i++) {
        totalDistance += graph[route[i]][route[i + 1]];
    }
    
    // Add distance from last city back to starting city
    totalDistance += graph[route[n - 1]][route[0]];
    
    return totalDistance;
}

/**
 * Generate all permutations and find minimum distance
 * @param graph Distance matrix
 * @param route Current route being built
 * @param visited Array to track visited cities
 * @param n Number of cities
 * @param level Current level in recursion
 * @param minDistance Pointer to minimum distance found so far
 * @param bestRoute Array to store the best route
 */
void tspBruteForce(int graph[][MAX_CITIES], int route[], int visited[], 
                   int n, int level, int* minDistance, int bestRoute[]) {
    // Base case: all cities visited
    if (level == n) {
        int currentDistance = calculateRouteDistance(graph, route, n);
        if (currentDistance < *minDistance) {
            *minDistance = currentDistance;
            // Copy current route to best route
            for (int i = 0; i < n; i++) {
                bestRoute[i] = route[i];
            }
        }
        return;
    }
    
    // Try all unvisited cities at current level
    for (int i = 0; i < n; i++) {
        if (!visited[i]) {
            visited[i] = 1;
            route[level] = i;
            
            // Recurse to next level
            tspBruteForce(graph, route, visited, n, level + 1, minDistance, bestRoute);
            
            // Backtrack
            visited[i] = 0;
        }
    }
}

/**
 * Solve TSP using brute force approach
 * @param graph Adjacency matrix representing distances between cities
 * @param n Number of cities
 * @param bestRoute Array to store the optimal route
 * @return Minimum distance for the shortest route
 */
int solveTSP(int graph[][MAX_CITIES], int n, int bestRoute[]) {
    if (n <= 1) {
        if (n == 1) bestRoute[0] = 0;
        return 0;
    }
    
    int route[MAX_CITIES];
    int visited[MAX_CITIES] = {0};
    int minDistance = INT_MAX;
    
    // Start from city 0
    visited[0] = 1;
    route[0] = 0;
    
    // Find optimal route
    tspBruteForce(graph, route, visited, n, 1, &minDistance, bestRoute);
    
    return minDistance;
}

/**
 * Helper function to print route
 */
void printRoute(int route[], int n) {
    printf("Route: ");
    for (int i = 0; i < n; i++) {
        printf("%d", route[i]);
        if (i < n - 1) printf(" -> ");
    }
    printf(" -> %d\n", route[0]); // Return to start
}

/**
 * Helper function to print distance matrix
 */
void printMatrix(int graph[][MAX_CITIES], int n) {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            printf("%3d ", graph[i][j]);
        }
        printf("\n");
    }
}

int main() {
    printf("=== Traveling Salesman Problem (TSP) - Brute Force ===\n");
    
    // Test Case 1: Simple 4-city problem
    printf("Test Case 1: 4 cities\n");
    int graph1[MAX_CITIES][MAX_CITIES] = {
        {0, 10, 15, 20},
        {10, 0, 35, 25},
        {15, 35, 0, 30},
        {20, 25, 30, 0}
    };
    int n1 = 4;
    int bestRoute1[MAX_CITIES];
    
    printf("Distance matrix:\n");
    printMatrix(graph1, n1);
    
    int minDistance1 = solveTSP(graph1, n1, bestRoute1);
    printf("Minimum distance: %d\n", minDistance1);
    printRoute(bestRoute1, n1);
    printf("\n");
    
    // Test Case 2: 3-city triangle
    printf("Test Case 2: 3 cities\n");
    int graph2[MAX_CITIES][MAX_CITIES] = {
        {0, 5, 8},
        {5, 0, 3},
        {8, 3, 0}
    };
    int n2 = 3;
    int bestRoute2[MAX_CITIES];
    
    printf("Distance matrix:\n");
    printMatrix(graph2, n2);
    
    int minDistance2 = solveTSP(graph2, n2, bestRoute2);
    printf("Minimum distance: %d\n", minDistance2);
    printRoute(bestRoute2, n2);
    printf("\n");
    
    // Test Case 3: 2-city problem (trivial)
    printf("Test Case 3: 2 cities\n");
    int graph3[MAX_CITIES][MAX_CITIES] = {
        {0, 7},
        {7, 0}
    };
    int n3 = 2;
    int bestRoute3[MAX_CITIES];
    
    printf("Distance matrix:\n");
    printMatrix(graph3, n3);
    
    int minDistance3 = solveTSP(graph3, n3, bestRoute3);
    printf("Minimum distance: %d\n", minDistance3);
    printRoute(bestRoute3, n3);
    printf("\n");
    
    // Test Case 4: Single city
    printf("Test Case 4: 1 city\n");
    int graph4[MAX_CITIES][MAX_CITIES] = {{0}};
    int n4 = 1;
    int bestRoute4[MAX_CITIES];
    
    printf("Distance matrix:\n");
    printMatrix(graph4, n4);
    
    int minDistance4 = solveTSP(graph4, n4, bestRoute4);
    printf("Minimum distance: %d\n", minDistance4);
    printRoute(bestRoute4, n4);
    
    return 0;
}